packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,196 @@
+      })
+    })
+
+    const { allowWrite, dispose: disposeReadonlyMW } = readonlyMiddleware(this.subtreeRootDraft)
+    this.allowWrite = allowWrite
+
+    this.disposer = () => {
+      disposeReactionRS()
+      disposeOnPatches()
+      disposeReadonlyMW()
+      if (isRootStore(this.subtreeRootDraft)) {
+        unregisterRootStore(this.subtreeRootDraft)
+      }
+      this.disposer = () => {}
+    }
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @typeparam R Return type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is called with a draft copy of `node`. Any changes made to the draft
+   * are applied to the original subtree when `fn` returns `true` or `{ commit: true, ... }`. When
+   * `fn` returns `false` or `{ commit: false, ... }` the changes made to the draft are rejected.
+   * @returns Value of type `R` when `fn` returns an object of type `{ commit: boolean; return: R }`
+   * or `void` when `fn` returns a boolean.
+   */
+  withDraft<T extends object, R = void>(node: T, fn: WithDraftCallback<T, R>): R {
+    assertIsFunction(fn, "fn")
+
+    const { draftNode, applyRecorderChanges } = this.prepareDraftChanges(node)
+
+    let commit = false
+    try {
+      const returnValue = this.allowWrite(() => fn(draftNode))
+      if (typeof returnValue === "boolean") {
+        commit = returnValue
+        return undefined as any
+      } else {
+        commit = returnValue.commit
+        return returnValue.return
+      }
+    } finally {
+      applyRecorderChanges(commit)
+    }
+  }
+
+  /**
+   * Disposes the draft middleware.
+   */
+  dispose(): void {
+    this.disposer()
+  }
+
+  private prepareDraftChanges<T extends object>(
+    node: T
+  ): { draftNode: T; applyRecorderChanges: (commit: boolean) => void } {
+    assertTweakedObject(node, "node")
+
+    const isDraft = !!this.withDraftPatchRecorder
+
+    const path = getParentToChildPath(isDraft ? this.subtreeRootDraft : this.subtreeRoot, node)
+    if (!path) {
+      throw failure(`node is not a child of subtreeRoot${isDraft ? "Draft" : ""}`)
+    }
+
+    const draftNode = resolvePath<T>(this.subtreeRootDraft, path).value
+    if (!draftNode) {
+      throw failure("path could not be resolved - draft may be out of sync with original tree")
+    }
+
+    if (!this.withDraftPatchRecorder) {
+      this.withDraftPatchRecorder = patchRecorder(this.subtreeRootDraft)
+    }
+    const recorder = this.withDraftPatchRecorder
+    const numRecorderEvents = recorder.events.length
+
+    const applyRecorderChanges = (commit: boolean): void => {
+      if (!isDraft) {
+        recorder.dispose()
+        this.withDraftPatchRecorder = undefined
+      }
+      runInAction(() => {
+        if (commit) {
+          if (!isDraft) {
```

Comment:
I'm torn about this.
On the one hand it might be confusing that the inner call does not actually commit anything, but on the other hand it would be weird for the inner call to commit changes it did not make.
Maybe the inner draft should internally generate another draft object (draft of the draft?), but that would be slow...

What are your thoughts on this?

Reply:
Ah, and I think this behavior (whichever is chosen) should be documented.

Reply:
> Maybe the inner draft should internally generate another draft object (draft of the draft?), but that would be slow...

Yes, ideally there should be no need to have a centrally managed draft, but cloning often is too slow. ...

> On the one hand it might be confusing that the inner call does not actually commit anything, but on the other hand it would be weird for the inner call to commit changes it did not make.

... So in my opinion this is an implementation detail that is necessary to support this API/DX. A developer shouldn't notice that committing changes from an inner call doesn't actually do anything because the changes have already been recorded due to the way this mechanism is implemented.

Or are you concerned about a userland `onPatches` listener that gets triggered slightly unexpectedly (not when the nested `withDraft` returns but immediately when the changes are made)?

Reply:
My concern is the duality when returning true:
- if it is un-nested it will commit changes right away
- if it is nested it will commit changes only if (and after) the outermost withDraft call returns true

I have the hunch that is the right behavior, but I also think it should be on the docs so it doesn't surprise anybody :)

Reply:
Okay, I'll add documentation about this behavior.